### DISCUSSION TITLE
locked AVR32DA28 - unlocking via prog.py

### DISCUSSION THREAD

**Discussion by flpxra** [SpenceKonde/DxCore#332] [Q&A] [answered]
I have a problem with unlocking AVR32DA28 after I set LOCK.KEY.

I set the lock by UPDI in this way:
`avrdude -c jtag2updi -p avr32da28 -P COM10 -U lock:w:0xa3:m`

I try to remove the lock by Serial UPDI and the script but it fails:
`DxCore\hardware\megaavr\1.4.10/tools/prog.py -t uart -u COM11 -b 57600 -d avr32da28 -a erase -v`

When using pymcuprog installed in a conventional non-portable Python installation, I can remove the lock with the command:
`pymcuprog erase  -t uart -u COM11 -d avr32da28 --chip-erase-locked-device`

From what I read in the `prog.py`  source code, it should be able to remove the lock every time a new code is uploaded to the MCU, but it doesn't work. I'm not a Python guy so I need help.

PS. prog.py looks very promising to me, it handles all dependecies of pymcuprog without installing it in conventional way. Also is there any method on prog.py to set lock on MCU via Serial UPDI?

Here is a output from console after trying to upload new code into a locked MCU:
```
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.2.3 - Jan 2022
Using serial port COM11 at 57600 baud.
Target: avr32da28
Set fuses: ['5:0b11001001', '6:0x04', '7:0x00', '8:0x00']
Action: write
File: C:\Users\domek\AppData\Local\Temp\arduino_build_383683/sketch_aug20a.ino.hex
pymcuprog.programmer - INFO - Setting up programming session for 'avr32da28'
pymcuprog.deviceinfo.deviceinfo - INFO - Looking for device avr32da28
pymcuprog.serialupdi.physical - INFO - Opening port 'COM11' at '57600' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - Setting UPDI clock to 4 MHz
pymcuprog.serialupdi.link - INFO - STCS 03 to 0x09
pymcuprog.serialupdi.physical - INFO - Switching to '57600' baud
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A3.KV00B.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A3.KV00B.0)'
pymcuprog.serialupdi.application - INFO - Using 24-bit UPDI
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x03
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - INFO - Entering NVM programming mode
pymcuprog.serialupdi.link - INFO - LDCS from 0x07
pymcuprog.serialupdi.application - INFO - Apply reset
pymcuprog.serialupdi.link - INFO - STCS 59 to 0x08
pymcuprog.serialupdi.application - INFO - Release reset
pymcuprog.serialupdi.link - INFO - STCS 00 to 0x08
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - ERROR - Timeout waiting for device to unlock
pymcuprog.nvm - ERROR - Device is locked.
Error:
Failed to enter NVM programming mode: device is locked
pymcuprog.nvm - INFO - No specific initializer for this provider
Pinging device...
pymcuprog.programmer - INFO - Reading device ID...
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A3.KV00B.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A3.KV00B.0)'
pymcuprog.serialupdi.application - INFO - Using 24-bit UPDI
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x03
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
Traceback (most recent call last):
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10/tools/prog.py", line 285, in <module>
    main()
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10/tools/prog.py", line 128, in main
    return_code = pymcuprog_basic(args, fuses_dict)
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10/tools/prog.py", line 216, in pymcuprog_basic
    pymcu._action_ping(backend)
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\pymcuprog_main.py", line 156, in _action_ping
    response = backend.read_device_id()
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\backend.py", line 399, in read_device_id
    return self.programmer.read_device_id()
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\programmer.py", line 130, in read_device_id
    return self.device_model.read_device_id()
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\nvmserialupdi.py", line 73, in read_device_id
    sig = self.avr.read_data(signatures_base, 3)
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\serialupdi\application.py", line 139, in read_data
    return self.readwrite.read_data(address, size)
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\serialupdi\readwrite.py", line 79, in read_data
    self.datalink.st_ptr(address)
  File "C:\Users\domek\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.4.10\tools\libs\pymcuprog\serialupdi\link.py", line 460, in st_ptr
    raise PymcuprogError("Error with st_ptr")
pymcuprog.pymcuprog_errors.PymcuprogError: Error with st_ptr
pymcuprog.serialupdi.physical - INFO - Closing port 'COM11'
the selected serial port pymcuprog.serialupdi.physical - INFO - Closing port 'COM11'
 does not exist or your board is not connected
```

**Comment by SpenceKonde** [comment]
We are already aware that prog.py does not support locking or unlocking chips.

Please be patient, I work 16 hours a day 7 days a week on either the cores or work for clients and the first person who even mentioned locking and unlocking with prog.py did so like a month ago  That's how long I have been aware that we never actually implemented it. Microchip has just released the DDs, and if you haven't noticed, the head of all three of my cores is completely busted and doesn't even compile blink - and I have to fix this schematic and layout for a client which eagle seems to have made absolutely impossible - all my attempts break forward and backward annotation (read "go back to your last backup and try again because you'll never fix it"). I'm on the fifth try right now. 

How about you try using whatever the hell you locked the chip with to unlock it? You didn't lock it with prog.py, because we don't support locking it either yet, so you must have another UPDI programmer, and it probably knows how to do a CHIPERASE to unlock the chip. 

Quentin, the coauthor of SerialUPDI - the useful one who actually knows python (Do you think I know python? It's only a slight exaggeration to say that stack overflow should be on the credits list ahead of me - you'll notice I've never once posted any code written for an atmega2560 - that's because my my brain only supports 16 bit address space (avr flash is word addressed) and the 17th bit would overflow the 16bit registers in my brain. I'm supposed to be able to turn two 16 bit mental registers into a 32-bit one so I could program ARM32 and x86 competently, but it doesn't seem to work right, and if you think the AVR DA errata listing is bad, you should check out the one for _homo sapiens_ (it's a bit of a download, though, and my model year was marked NRND so I have little hope for an upgrade there. 

Anyway, he appears to be a more advanced model and understands python reasonably well, so we're hoping to hammer out a number of issues with serialUPDI during that time including this since we will be able to co-work in person. 

Now, if it's not fixed by thanksgiving, that's when you start thinking of ways to make him miss his flight home.

**Comment by flpxra** [comment, marked as answer]
> How about you try using whatever the hell you locked the chip with to unlock it? You didn't lock it with prog.py, because we don't support locking it either yet, so you must have another UPDI programmer, and it probably knows how to do a CHIPERASE to unlock the chip.

If I could, I would have done it, but of course: it doesn't work. I started an issue in the `jtag2updi` project.

Temporarily I used portable `WinPython` with `pymcuprog` and now I can lock and unlock MCU via Serial UPDI.

**Comment by SpenceKonde** [comment]
Yeah, jtag2updi when I was going through hell working on it because of the undocumented behavior of UPDI on 24-bit addressed devices, (took me two weeks to figure out that a direct write to 0x1000 meant something different before and after writing to a 24 bit address). I was just going mad, it did chip erase by entering nvm prog mode and programming that way (also I was just looking at his current code, and it's very clear that he knows the on-chip debugging key... but it's not in the public repo. It was leaked once in an early datasheet, but I can't find a copy of it and I am very annoyed by that). 

Who was the person who wanted to make a debugger for these parts though? jtag2updi has all the debug commands in it. would be very useful to whoever was taking on that nightmare of a task, then again his code is so obtuse it might not be so useful after all. And it is not valuable to have at would take seconds to extract the ocd key from soemone using an official microchip tool, that's not the barrier to OCD, the barrier is that you need to interface it with the software on the host. 

Yeah, I think until we sort out prog,py unlock that's your best bet - though you can probably use the version of python that we bundle. I might have code that isn't yet in head that can unlock, but not sure if it works

**Comment by X0NARU** [comment]
> I set the lock by UPDI in this way:
> `avrdude -c jtag2updi -p avr32da28 -P COM10 -U lock:w:0xa3:m`


Can you please explain this command `lock:w:0xa3:m` , I need to lock avd32dd20 , how I need to modify it ?
